internal/ethapi/api.go

```diff
@@ -1627,9 +1627,14 @@ func AccessList(ctx context.Context, b Backend, blockNrOrHash rpc.BlockNumberOrH
 		return nil, 0, nil, err
 	}
 
-	// Ensure any missing fields are filled, extract the recipient and input data
-	if err := args.setDefaults(ctx, b, true); err != nil {
-		return nil, 0, nil, err
+	// fill in missing fields based on the state at the specified block
+	if args.Nonce == nil {
+		nonce := hexutil.Uint64(db.GetNonce(args.from()))
+		args.Nonce = &nonce
+	}
+	blockCtx := core.NewEVMBlockContext(header, NewChainContext(ctx, b), nil)
+	if err := args.CallDefaults(b.RPCGasCap(), blockCtx.BaseFee, b.ChainConfig().ChainID); err != nil {
```

Comment:
the `setDefaults` has a lot of logic handling blobs that are passed in the args, and that will become broken if we just change this

Reply:
we could just call `setBlobSidecar` in CallDefaults

Reply:
But `setDefaults` will set the fee defaults on the latest block. 
``
	if err := args.setFeeDefaults(ctx, b); err != nil {
		return err
	}
```
And regardless of what order we invoke it, it will use the latest header, and do 
```
	if err := args.setCancunFeeDefaults(ctx, head, b); err != nil {
		return err
	}
```
SO it will apply cancun checks according to `tip`, regardless of what the given block is.

Reply:
How about 
```golang
// setFeeDefaults fills in default fee values for unspecified tx fields.
func (args *TransactionArgs) setFeeDefaults(ctx context.Context, b Backend) error {
	return setFeeDefaultsUsingThisHeader(b.CurrentHeader())
}
func (args *TransactionArgs) setFeeDefaultsUsingThisHeader(ctx context.Context, b Backend, head *types.Header) error {
	// Sanity check the EIP-4844 fee parameters.
```

Reply:
> the setDefaults has a lot of logic handling blobs that are passed in the args, and that will become broken if we just change this

The logic there is to fill in blob proofs and commitments if they were not provided. Those matter only when submitting a tx. Since for a simulation all that can be accessed via evm is blob hashes (which still can be passed in via TransactionArgs).

All in all I think I agree with @jwasinger that CallDefaults is a better validation function to use here. All the methods that involve signing and submitting a tx use `setDefaults` while all the "simulation" methods use CallDefaults.